### PR TITLE
lib/master: implement and use ErrCenter

### PR DESCRIPTION
**Pull request by liuzix**
- Explore a more managed mechanism of error handling inside `BaseMaster` and `BaseWorker`.

Abstractly speaking, this model is like this: there is an object O, and there are methods A, B, C, D on it. These methods may be called by many goroutines. These methods may depend on other modules or third-party libraries, and the dependent methods may block. Suppose an error  occurs in a method of object O, and we want to stop all methods of O as soon as possible, and call it again with an error.

### CODE REVIEW

File: pkg/errctx/center.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package errctx
+
+import (
+	"context"
+
+	"go.uber.org/atomic"
+)
+
+type ErrCenter struct {
```
Comment (lance6716) [inline pkg/errctx/center.go:9]: prefer the name `FirstError`

File: pkg/errctx/center.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	}
+	c.errVal.Store(err)
+	close(c.doneCh)
+}
+
+func (c *ErrCenter) CheckError() error {
+	return c.errVal.Load()
+}
+
+func (c *ErrCenter) DeriveContext(ctx context.Context) context.Context {
```
Comment (lance6716) [inline pkg/errctx/center.go:38]: we can make it more like a standard library stype, WithCancelOnError(ctx, firstErr) or something

File: pkg/errctx/ctx.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	// has true.
+	c.doneCh = doneCh
+	c.hasDoneCh.Store(true)
+
+	return doneCh
+}
+
+func (c *errCtx) Err() error {
+	if c.center.hasErr.Load() {
+		return c.center.errVal.Load()
```
Comment (lance6716) [inline pkg/errctx/ctx.go:63]: there's a time slice in setting `hasErr` and `errVal` in `OnError`

Comment (liuzix) [reply pkg/errctx/ctx.go:63]: Good catch. I will use an RW-lock instead.

File: pkg/errctx/center.go (hunk 1 above)
Comment (liuzix) [reply pkg/errctx/center.go:9]: I think `FirstError` is too simple to describe the object's functionality. Don't you think so?

Comment (lance6716) [reply pkg/errctx/center.go:9]: From `ErrCenter` I can not know its functionality, maybe left some comments

File: pkg/errctx/center.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+func (c *ErrCenter) OnError(err error) {
+	if err == nil {
+		return
+	}
+	if c.hasErr.Swap(true) {
+		// OnError is no-op after the first call with
+		// a non-nil error.
+		return
+	}
```
Comment (amyangfei) [inline pkg/errctx/center.go:29]: If some of scenarios need to record the duplicated errors, should we provide an option to do that.

File: pkg/errctx/ctx.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func (c *errCtx) Done() <-chan struct{} {
+	if c.hasDoneCh.Load() {
+		// Fast path. No lock contention.
+		return c.doneCh
+	}
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
```
Comment (maxshuang) [inline pkg/errctx/ctx.go:33]: sync.Once may be a better choice ??

File: pkg/errctx/center.go (hunk 4 above)
Comment (liuzix) [reply pkg/errctx/center.go]: Allowing duplicate errors can make this utility too difficult to use. I added a log instead.

File: pkg/errctx/ctx.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+			case <-c.center.doneCh:
+			case <-c.Context.Done():
+			}
+
+			close(doneCh)
+		}()
+
+		// There is no data race here because
+		// c.doneCh is read from only if c.hasDoneCh
+		// has true.
```
Comment (amyangfei) [inline pkg/errctx/ctx.go:38]: The comment is outdated

**Review by amyangfei — COMMENTED** [review]
rest lgtm

**Review by amyangfei — APPROVED** [review]

**Pull request merged**